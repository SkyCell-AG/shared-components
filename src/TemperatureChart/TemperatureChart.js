import React, {
    useMemo,
    useCallback,
} from 'react'
import PropTypes from 'prop-types'
import {
    VictoryLine,
    VictoryChart,
    VictoryAxis,
    LineSegment,
} from 'victory'

import getScale from './getScaleOffset'

const propTypes = {
    ambient: PropTypes.arrayOf(PropTypes.number),
    simulated: PropTypes.arrayOf(PropTypes.number),
    energyLevel: PropTypes.arrayOf(PropTypes.number),
    upperTempBound: PropTypes.number,
    lowerTempBound: PropTypes.number,
    excursion: PropTypes.string,
    temperatureTimeAxis: PropTypes.arrayOf(PropTypes.string),
    style: PropTypes.shape({
        axisLineStyle: PropTypes.shape(),
        rangeLineStyle: PropTypes.shape(),
        simulated: PropTypes.shape(),
        ambient: PropTypes.shape(),
        energyLevel: PropTypes.shape(),
        excursionAxis: PropTypes.shape(),
    }),
    width: PropTypes.number.isRequired,
}

const defaultProps = {
    ambient: undefined,
    simulated: undefined,
    energyLevel: undefined,
    upperTempBound: null,
    lowerTempBound: null,
    excursion: undefined,
    temperatureTimeAxis: [],
    style: {
        axisLineStyle: undefined,
        simulated: undefined,
        ambient: undefined,
        energyLevel: undefined,
        rangeLineStyle: undefined,
        excursionAxis: undefined,
    },
}

const grey = '#00000066'

const axisStyle = {
    axisLabel: {
        fontSize: 8,
        padding: 30,
        fill: grey,
        fontFamily: 'Arimo, Roboto, Arial, sans-serif',
    },
    tickLabels: {
        fontSize: 8,
        padding: 2,
        fontFamily: 'Arimo, Roboto, Arial, sans-serif',
        fill: grey,
    },
}

const axisLineStyle = {
    stroke: grey,
}

const generateBorderLabel = (title) => {
    return ({
        datum,
        data,
    }) => {
        if (datum === data[data.length - 1]) {
            return title
        }

        return ''
    }
}

const rangeLineStyle = {
    data: {
        stroke: '#edae49',
        strokeDasharray: [
            3,
            2,
        ],
        strokeWidth: 1.5,
    },
    parent: {
        border: `1px solid ${grey}`,
    },
    labels: {
        fontSize: 8,
        padding: 2,
        fill: '#edae49',
    },
}

const TemperatureChart = ({
    ambient,
    simulated,
    energyLevel,
    upperTempBound,
    lowerTempBound,
    excursion,
    style,
    temperatureTimeAxis,
    width,
}) => {
    const scale = useMemo(() => {
        return getScale({
            ambient,
            simulated,
        })
    }, [
        ambient,
        simulated,
    ])

    const energyLevelScaled = useMemo(() => {
        if (!energyLevel) {
            return undefined
        }

        return energyLevel.map((val) => { return val * scale })
    }, [
        scale,
        energyLevel,
    ])

    const tickPercentValues = useMemo(() => {
        return [
            0,
            20,
            40,
            60,
            80,
            100,
        ].map((val) => { return val * scale })
    }, [scale])

    const tickPercentFormat = useCallback((val) => {
        return `${Math.round(val / scale)}%`
    }, [scale])

    const tickTemperatureTimeValues = useMemo(() => {
        return temperatureTimeAxis.map((element) => {
            const date = new Date(element)

            return `${date.getDate()}/${date.getMonth()}/${date.getFullYear()} ${date.getHours()}:${date.getMinutes()}`
        })
    }, [temperatureTimeAxis])

    const tickTemperatureTimeFormat = useCallback((t) => {
        return t
    }, [])

    const excursionIndex = useMemo(() => {
        return temperatureTimeAxis.findIndex((element) => {
            return element === excursion
        }) + 1
    }, [
        temperatureTimeAxis,
        excursion,
    ])

    const generateBoundaryData = useCallback((boundary) => {
        return [
            {
                x: 0,
                y: boundary,
            },
            {
                x: simulated.length - 1,
                y: boundary,
            },
        ]
    }, [simulated])

    const tempRangeTopData = useMemo(() => {
        return generateBoundaryData(upperTempBound)
    }, [
        generateBoundaryData,
        upperTempBound,
    ])

    const tempRangeBottomData = useMemo(() => {
        return generateBoundaryData(lowerTempBound)
    }, [
        generateBoundaryData,
        lowerTempBound,
    ])

    const getTemperaturePositions = useCallback((temperature) => {
        return temperature.map((element, i) => {
            return {
                x: i + 1,
                y: element,
            }
        })
    }, [])

    const ambientTemperature = useMemo(() => {
        return getTemperaturePositions(ambient)
    }, [
        getTemperaturePositions,
        ambient,
    ])

    const simulatedTemperature = useMemo(() => {
        return getTemperaturePositions(simulated)
    }, [
        getTemperaturePositions,
        simulated,
    ])

    return (
        <VictoryChart width={width}>
            <VictoryAxis
                data-testid="temperatureTimeAxis"
                style={axisStyle}
                gridComponent={(
                    <LineSegment
                        style={axisLineStyle}
                    />
                )}
                domain={[
                    1,
                    temperatureTimeAxis.length - 1,
                ]}
                tickValues={tickTemperatureTimeValues}
                tickFormat={tickTemperatureTimeFormat}
            />
            <VictoryAxis style={{
                ...axisStyle,
                axis: {
                    stroke: grey,
                    strokeWidth: 1,
                },
            }}
            />
            <VictoryAxis
                dependentAxis
                style={{
                    ...axisStyle,
                    axis: {
                        stroke: 'transparent',
                    },
                }}
                gridComponent={(
                    <LineSegment
                        style={style.axisLineStyle || axisLineStyle}
                    />
                )}
            />
            {energyLevel && (
                <VictoryAxis
                    data-testid="energyLevelAxis"
                    dependentAxis
                    orientation="right"
                    label="Energy Level"
                    tickValues={tickPercentValues}
                    tickFormat={tickPercentFormat}
                    style={axisStyle}
                    gridComponent={(
                        <LineSegment
                            style={axisLineStyle}
                        />
                    )}
                />
            )}
            <VictoryLine
                data={simulatedTemperature}
                style={style.simulated || {
                    data: {
                        stroke: '#61c6e9',
                        strokeWidth: 1.5,
                    },
                }}
            />
            {energyLevel && (
                <VictoryLine
                    data-testid="energyLevelLine"
                    data={energyLevelScaled}
                    style={style.energyLevel || {
                        data: {
                            stroke: '#9e9e9e',
                            strokeWidth: 1.5,
                        },
                    }}
                />
            )}
            <VictoryLine
                data={ambientTemperature}
                style={style.ambient || {
                    data: {
                        stroke: '#cf3b8a',
                        strokeWidth: 1.5,
                    },
                }}
            />
            <VictoryLine
                data={tempRangeTopData}
                style={style.rangeLineStyle || rangeLineStyle}
                labels={generateBorderLabel('Max')}
            />
            <VictoryLine
                data={tempRangeBottomData}
                style={style.rangeLineStyle || rangeLineStyle}
                labels={generateBorderLabel('Min')}
            />
            {excursion !== undefined && (
                <VictoryAxis
                    data-testid="excursionAxis"
                    dependentAxis
                    label="Excursion"
                    axisValue={excursionIndex}
                    style={style.excursionAxis || {
                        axis: {
                            stroke: 'red',
                        },
                        tickLabels: {
                            fill: 'none',
                        },
                        axisLabel: {
                            fontSize: 8,
                            padding: -10,
                        },
                    }}
                />
            )}
        </VictoryChart>
    )
}

TemperatureChart.propTypes = propTypes
TemperatureChart.defaultProps = defaultProps

export default TemperatureChart
