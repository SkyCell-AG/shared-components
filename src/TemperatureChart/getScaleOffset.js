const getScaleOffset = ({
    ambient,
    simulated,
}) => {
    const max = Math.max(...ambient, ...simulated)
    const min = Math.min(...ambient, ...simulated)

    const range = max - min
    const scale = range / 100

    return scale
}

export default getScaleOffset
