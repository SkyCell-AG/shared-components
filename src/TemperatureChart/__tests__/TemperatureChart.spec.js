import React from 'react'
import {
    shallow,
} from 'enzyme'

import {
    VictoryAxis,
} from 'victory'

import TemperatureChart from '../TemperatureChart'

describe('TemperatureChart', () => {
    it('TemperatureChart success', () => {
        const wrapper = shallow(
            <TemperatureChart
                containerSensorData={{
                    data: [
                        {
                            d: [
                                2,
                                20,
                            ],
                            t: '03.12.2021 09:10:00',
                        },
                        {
                            d: [
                                3,
                                52,
                            ],
                            t: '03.12.2021 09:20:00',

                        },
                        {
                            d: [
                                1,
                                32,
                            ],
                            t: '03.12.2021 09:30:00',
                        },
                    ],
                }}
                energyLevel={[
                    100,
                    98,
                    96,
                ]}
                excursion="2021-12-03T09:20"
                upperTempBound={8}
                lowerTempBound={2}
                width={1300}
            />,
        )

        expect(wrapper).toMatchSnapshot()
    })

    it('Should not render energyLevel if energyLevel undefined', () => {
        const wrapper = shallow(
            <TemperatureChart
                containerSensorData={{
                    data: [
                        {
                            d: [
                                2,
                                20,
                            ],
                            t: '03.12.2021 09:10:00',
                        },
                        {
                            d: [
                                3,
                                52,
                            ],
                            t: '03.12.2021 09:20:00',

                        },
                        {
                            d: [
                                1,
                                32,
                            ],
                            t: '03.12.2021 09:30:00',
                        },
                    ],
                }}
                excursion="2021-12-03T09:20"
                upperTempBound={8}
                lowerTempBound={2}
                width={1300}
            />,
        )

        expect(wrapper.find('[data-testid="energyLevelAxis"]').exists()).toEqual(false)
        expect(wrapper.find('[data-testid="energyLevelLine"]').exists()).toEqual(false)
    })

    it('Should not render excursion if excursion undefined', () => {
        const wrapper = shallow(
            <TemperatureChart
                containerSensorData={{
                    data: [
                        {
                            d: [
                                2,
                                20,
                            ],
                            t: '03.12.2021 09:10:00',
                        },
                        {
                            d: [
                                3,
                                52,
                            ],
                            t: '03.12.2021 09:20:00',

                        },
                        {
                            d: [
                                1,
                                32,
                            ],
                            t: '03.12.2021 09:30:00',
                        },
                    ],
                }}
                upperTempBound={8}
                lowerTempBound={2}
                width={1300}
            />,
        )

        expect(wrapper.find('[data-testid="excursionAxis"]').exists()).toEqual(false)
    })

    it('tickFormat function returns the right data', () => {
        const wrapper = shallow(
            <TemperatureChart
                containerSensorData={{
                    data: [
                        {
                            d: [
                                2,
                                20,
                            ],
                            t: '03.12.2021 09:10:00',
                        },
                        {
                            d: [
                                3,
                                52,
                            ],
                            t: '03.12.2021 09:20:00',

                        },
                        {
                            d: [
                                1,
                                32,
                            ],
                            t: '03.12.2021 09:30:00',
                        },
                    ],
                }}
                energyLevel={[
                    100,
                    98,
                    96,
                ]}
                excursion="2021-12-03T09:20"
                upperTempBound={8}
                lowerTempBound={2}
                width={1300}
            />,
        )

        const child = wrapper.find(VictoryAxis)

        expect(child.at(0).props().tickFormat('3/11/2021 9:10')).toEqual('3/11/2021 9:10')
    })
})
