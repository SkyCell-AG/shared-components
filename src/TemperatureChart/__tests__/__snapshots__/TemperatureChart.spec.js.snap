// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemperatureChart success 1`] = `
<VictoryChart>
  <VictoryAxis
    style={
      Object {
        "axis": Object {
          "stroke": "#00000066",
          "strokeWidth": 1,
        },
        "axisLabel": Object {
          "fill": "#00000066",
          "fontFamily": "Arimo, Roboto, Arial, sans-serif",
          "fontSize": 8,
          "padding": 30,
        },
        "tickLabels": Object {
          "fill": "#00000066",
          "fontFamily": "Arimo, Roboto, Arial, sans-serif",
          "fontSize": 8,
          "padding": 2,
        },
      }
    }
  />
  <VictoryLabel
    style={
      Object {
        "fill": "#00000066",
        "fontFamily": "Arimo, Roboto, Arial, sans-serif",
        "fontSize": 8,
        "padding": 30,
      }
    }
    text="Temperature"
    x={15}
    y={55}
  />
  <VictoryAxis
    dependentAxis={true}
    gridComponent={
      <LineSegment
        style={
          Object {
            "stroke": "#00000066",
          }
        }
      />
    }
    style={
      Object {
        "axis": Object {
          "stroke": "transparent",
        },
        "axisLabel": Object {
          "fill": "#00000066",
          "fontFamily": "Arimo, Roboto, Arial, sans-serif",
          "fontSize": 8,
          "padding": 30,
        },
        "tickLabels": Object {
          "fill": "#00000066",
          "fontFamily": "Arimo, Roboto, Arial, sans-serif",
          "fontSize": 8,
          "padding": 2,
        },
      }
    }
  />
  <VictoryAxis
    dependentAxis={true}
    gridComponent={
      <LineSegment
        style={
          Object {
            "stroke": "#00000066",
          }
        }
      />
    }
    label="Energy Level"
    orientation="right"
    style={
      Object {
        "axisLabel": Object {
          "fill": "#00000066",
          "fontFamily": "Arimo, Roboto, Arial, sans-serif",
          "fontSize": 8,
          "padding": 30,
        },
        "tickLabels": Object {
          "fill": "#00000066",
          "fontFamily": "Arimo, Roboto, Arial, sans-serif",
          "fontSize": 8,
          "padding": 2,
        },
      }
    }
    tickFormat={[Function]}
    tickValues={
      Array [
        0,
        10.2,
        20.4,
        30.6,
        40.8,
        51,
      ]
    }
  />
  <VictoryLine
    data={
      Array [
        2,
        3,
        1,
      ]
    }
    style={
      Object {
        "data": Object {
          "stroke": "#61c6e9",
          "strokeWidth": 1.5,
        },
      }
    }
  />
  <VictoryLine
    data={
      Array [
        51,
        49.980000000000004,
        48.96,
      ]
    }
    style={
      Object {
        "data": Object {
          "stroke": "#9e9e9e",
          "strokeWidth": 1.5,
        },
      }
    }
  />
  <VictoryLine
    data={
      Array [
        20,
        52,
        32,
      ]
    }
    style={
      Object {
        "data": Object {
          "stroke": "#cf3b8a",
          "strokeWidth": 1.5,
        },
      }
    }
  />
  <VictoryLine
    data={
      Array [
        Object {
          "x": 0,
          "y": 8,
        },
        Object {
          "x": 2,
          "y": 8,
        },
      ]
    }
    labels={[Function]}
    style={
      Object {
        "data": Object {
          "stroke": "#edae49",
          "strokeDasharray": Array [
            3,
            2,
          ],
          "strokeWidth": 1.5,
        },
        "labels": Object {
          "fill": "#edae49",
          "fontSize": 8,
          "padding": 2,
        },
        "parent": Object {
          "border": "1px solid #00000066",
        },
      }
    }
  />
  <VictoryLine
    data={
      Array [
        Object {
          "x": 0,
          "y": 2,
        },
        Object {
          "x": 2,
          "y": 2,
        },
      ]
    }
    labels={[Function]}
    style={
      Object {
        "data": Object {
          "stroke": "#edae49",
          "strokeDasharray": Array [
            3,
            2,
          ],
          "strokeWidth": 1.5,
        },
        "labels": Object {
          "fill": "#edae49",
          "fontSize": 8,
          "padding": 2,
        },
        "parent": Object {
          "border": "1px solid #00000066",
        },
      }
    }
  />
  <VictoryAxis
    axisValue={4}
    dependentAxis={true}
    label="Excursion"
    style={
      Object {
        "axis": Object {
          "stroke": "red",
        },
        "axisLabel": Object {
          "fontSize": 8,
          "padding": -10,
        },
        "tickLabels": Object {
          "fill": "none",
        },
      }
    }
  />
</VictoryChart>
`;
